File: src/main/java/org/example/tree/MaxDepthTree.java
package org.example.tree;

public class MaxDepthTree {

    public int maxDepth(TreeNode root) {
        if (root == null) return 0;

        return calcDepth(root, 1);
    }

    private int calcDepth(TreeNode node, int depth) {
        if (node == null || (node.left == null && node.right == null)) return depth;

        int leftDepth = 0, rightDepth = 0;
        if (node.left != null) {
            leftDepth = Math.max(depth, calcDepth(node.left, depth + 1));
        }
        if (node.right != null) {
            rightDepth = Math.max(depth, calcDepth(node.right, depth + 1));
        }

        return Math.max(leftDepth, rightDepth);
    }

    public static void main(String[] args) {
        TreeNode node7 = new TreeNode(7, null, null);
        TreeNode node15 = new TreeNode(15, null, null);
        TreeNode node20 = new TreeNode(20, node15, node7);
        TreeNode node9 =  new TreeNode(9, null, null);
        TreeNode node3 = new TreeNode(3, node9, node20);

        System.out.println(new MaxDepthTree().maxDepth(node3));
    }
}

File: src/main/java/org/example/tree/MaxDepthTree.py
from typing import Optional

from org.example.tree.TreeNode import TreeNode


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0

        def dfs(d, node):
            if not node or (not node.left and not node.right): return d
            leftDepth = rightDepth = 0
            if node.left:
                leftDepth = max(d, dfs(d + 1, node.left))
            if node.right:
                rightDepth = max(d, dfs(d + 1, node.right))
            return max(leftDepth, rightDepth)

        depth = dfs(1, root)
        return depth
